text/0000-ppv.md

```diff
@@ -0,0 +1,1596 @@
+extend Rust with a total order for floats should extend the portable
+floating-point vector types with it as well. See [this internal
+thread](https://users.rust-lang.org/t/how-to-sort-a-vec-of-floats/2838/3) for
+more information.
+
+### Inherent Methods
+
+#### Construction and element access
+
+All portable signed integer, unsigned integer, and floating-point vector types
+implement the following methods:
+
+```rust
+impl {element_type}{lane_width}x{number_of_lanes} {
+/// Creates a new instance of the vector from `number_of_lanes` 
+/// values.
+pub const fn new(args...: element_type) -> Self;
+
+/// Returns the number of vector lanes.
+pub const fn lanes() -> usize;
+
+/// Constructs a new instance with each element initialized to
+/// `value`.
+pub const fn splat(value: element_type) -> Self;
+
+/// Extracts the value at `index`.
+///
+/// # Panics
+///
+/// If `index >= Self::lanes()`.
+pub fn extract(self, index: usize) -> element_type;
+
+/// Extracts the value at `index`.
+///
+/// If `index >= Self::lanes()` the behavior is undefined.
+pub unsafe fn extract_unchecked(self, index: usize) -> element_type;
+
+/// Returns a new vector where the value at `index` is replaced by `new_value`.
+///
+/// # Panics
+///
+/// If `index >= Self::lanes()`.
+#[must_use = error-message]
+pub fn replace(self, index: usize, new_value: $elem_ty) -> Self;
+
+/// Returns a new vector where the value at `index` is replaced by `new_value`.
+#[must_use = error-message]
+pub unsafe fn replace_unchecked(self, index: usize, 
+                                new_value: element_type) -> Self;
+}
+```
+
+#### Reads and Writes
+
+##### Contiguous reads and writes
+
+All portable vector types implement the following methods:
+
+```rust
+impl {element_type}{lane_width}x{number_of_lanes} {
+/// Writes the values of the vector to the `slice` without 
+/// reading or dropping the old value.
+///
+/// # Panics
+///
+/// If `slice.len() != Self::lanes()` or `&slice[0]` is not
+/// aligned to an `align_of::<Self>()` boundary.
+pub fn write_aligned(self, slice: &mut [element_type]);
+
+/// Writes the values of the vector to the `slice` without 
+/// reading or dropping the old value.
+///
+/// # Panics
+///
+/// If `slice.len() != Self::lanes()`.
+pub fn write_unaligned(self, slice: &mut [element_type]);
+
+/// Writes the values of the vector to the `slice` without 
+/// reading or dropping the old value.
+///
+/// # Precondition
+///
+/// If `slice.len() < Self::lanes()` or `&slice[0]` is not
+/// aligned to an `align_of::<Self>()` boundary, the behavior is
+/// undefined.
+pub unsafe fn write_aligned_unchecked(self, slice: &mut [element_type]);
+
+/// Writes the values of the vector to the `slice` without reading 
+/// or dropping the old value.
+///
+/// # Precondition
+///
+/// If `slice.len() < Self::lanes()` the behavior is undefined.
+pub unsafe fn write_unaligned_unchecked(self, slice: &mut [element_type]);
```

Comment:
I was going to make the same comment here (and `read_unaligned_unchecked`) as for the aligned variants, only referring to `ptr::{read,write}_unaligned` instead. Then I realized this might be intended to still require being aligned for the element type? If so, that seems more useful, but

1. this subtlety should be called out explicitly
2. it still duplicates `ptr::write::<[element_type; Self::lanes()]>`, doesn't it?
3. same issue with uninitialized memory as in `write_aligned`

Reply:
I think that here we should do exactly the same thing we do for arrays with a `#[repr(align(N))]` attribute since these are in practice equivalent to vectors.  So given:

```rust
#[repr(align(N))]
struct V([E; M]);
```

where `align_of::<N>() > align_of::<[E; M]>()`. AFAIK `ptr::write_unaligned::<V>` allows me to write `V` to a pointer with alignment < `N`. 

Does it allow the `E`s to be at addresses with alignment < `align_of::<E>()` ?

Reply:
`ptr::{read,write}_unaligned` always lower the alignment requirement to 1. So would not assume that the elements are aligned to `align_of::<E>()`. This is not possible with an interface taking `&mut [E]` or `&mut E` (since references must always be aligned).

However, it seems theoretically useful to have a way to read and write vectors that aren't naturally aligned, but where the elements are aligned. Still, as I said, you could achieve that by doing *aligned* accesses via arrays (arrays have the same alignment requirements as their element type) + type punning, so I don't think this API is needed to achieve that.

Reply:
Technically, none of the APIs here are needed, since they are all implemented on top of `ptr::{read,write}{_unaligned}`. Their purpose is not to allow performing all the operations that the `ptr` methods allow, but to simplify reading and writing the portable vector types from/to slices. 

The advantages for the safe methods are more clear because they are safe as opposed to the pointer methods. However, the checks that make them safe do incur a cost that is often not acceptable when writing SIMD code. The `_unchecked` variants provide an easy way to disable these checks without having to fall back to using the raw `ptr` APIs. 

Also, even while one is dealing with raw pointers, it is still often more useful to create an slice from them if the memory is initialized and properly aligned, and then use these APIs, then doing the type punning required to map the portable vector types to arrays, casting that back to pointers, and so on.